py_ecc/fields/optimized_field_elements.py

```diff
@@ -0,0 +1,395 @@
+                int(c) * other % self.field_modulus
+                for c
+                in self.coeffs
+            ], self.curve_name)
+        elif isinstance(other, FQP):
+            b = [0] * (self.degree * 2 - 1)
+            inner_enumerate = list(enumerate(other.coeffs))
+            for i, eli in enumerate(self.coeffs):
+                for j, elj in inner_enumerate:
+                    b[i + j] += int(eli * elj)
+            # MID = len(self.coeffs) // 2
+            for exp in range(self.degree - 2, -1, -1):
+                top = b.pop()
+                for i, c in self.mc_tuples:
+                    b[exp + i] -= top * c
+            return type(self)([x % self.field_modulus for x in b], self.curve_name)
+        else:
+            raise ValueError(
+                "Expected an int or FQP object, but got object of type {}"
+                .format(type(other))
+            )
+
+    def __rmul__(self, other: Union[int, "FQP"]) -> "FQP":
+        return self * other
+
+    def __div__(self, other: Union[int, "FQ", "FQP"]) -> "FQP":
+        if isinstance(other, int):
+            return type(self)([
+                int(c) * prime_field_inv(other, self.field_modulus) % self.field_modulus
+                for c
+                in self.coeffs
+            ], self.curve_name)
+        elif isinstance(other, type(self)):
+            return self * other.inv()
+        else:
+            raise ValueError(
+                "Expected an int or FQP object, but got object of type {}"
+                .format(type(other))
+            )
+
+    def __truediv__(self, other: Union[int, "FQ", "FQP"]) -> "FQP":
+        return self.__div__(other)
+
+    def __pow__(self, other: int) -> "FQP":
+        o = type(self)([1] + [0] * (self.degree - 1), self.curve_name)
+        t = self
+        while other > 0:
+            if other & 1:
+                o = o * t
+            other >>= 1
+            t = t * t
+        return o
+
+    def optimized_poly_rounded_div(self,
+                                   a: Sequence[IntOrFQ],
+                                   b: Sequence[IntOrFQ]) -> Sequence[IntOrFQ]:
+        dega = deg(a)
+        degb = deg(b)
+        temp = [x for x in a]
+        o = [0 for x in a]
+        for i in range(dega - degb, -1, -1):
+            o[i] = int(o[i] + temp[degb + i] * prime_field_inv(int(b[degb]), self.field_modulus))
+            for c in range(degb + 1):
+                temp[c + i] = (temp[c + i] - o[c])
+        return [x % self.field_modulus for x in o[:deg(o) + 1]]
+
+    # Extended euclidean algorithm used to find the modular inverse
+    def inv(self) -> "FQP":
+        lm, hm = [1] + [0] * self.degree, [0] * (self.degree + 1)
+        low, high = (
+            cast(List[IntOrFQ], list(self.coeffs + (0,))),
+            cast(List[IntOrFQ], list(self.modulus_coeffs + (1,))),
+        )
+        low, high = list(self.coeffs + (0,)), self.modulus_coeffs + (1,)  # type: ignore
```

Comment:
This line appears to be redundant, can be deleted. (I see it was in the original)